libs/ml/src/core/node.cpp

```diff
@@ -0,0 +1,223 @@
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "ml/core/node.hpp"
+#include "ml/core/subgraph.hpp"
+
+namespace fetch {
+namespace ml {
+
+using Shape       = fetch::math::SizeVector;
+using ShapeVector = std::vector<fetch::math::SizeVector>;
+
+/**
+ * returns the stored operation type and syncs it with operation type of
+ * underlying Ops.
+ * @return
+ */
+template <typename TensorType>
+OpType Node<TensorType>::OperationType()
+{
+  if (operation_type_ != op_ptr_->OperationType())
+  {
+    FETCH_LOG_ERROR(this->name_.c_str(), "Node operation type (" +
+                                             std::to_string(static_cast<int>(operation_type_)) +
+                                             ") and underlying Ops operation code (" +
+                                             std::string(op_ptr_->Descriptor()) + ") mismatch!");
+    operation_type_ = op_ptr_->OperationType();
+  }
+  return operation_type_;
+}
+
+template <typename TensorType>
+bool Node<TensorType>::HasValidCache()
+{
+  return static_cast<bool>(cached_output_status_ == CachedOutputState::VALID_CACHE);
+}
+
+template <typename TensorType>
+void Node<TensorType>::SetBatchOutputShape(const Shape &new_shape)
+{
+  op_ptr_->SetBatchOutputShape(new_shape);
+}
+
+template <typename TensorType>
+void Node<TensorType>::SetBatchInputShapes(const ShapeVector &new_shapes)
+{
+  op_ptr_->SetBatchInputShapes(new_shapes);
+}
+
+template <typename TensorType>
+const ShapeVector &Node<TensorType>::BatchInputShapes()
+{
+  return op_ptr_->BatchInputShapes();
+}
+
+/**
+ * @brief BatchOutputShape computes an output shape of the Node, if only 1 data slice (e.g.
+ * with batch size == 1) is provided to the Graph input. If there is no cached output shape,
+ * the method is recursively called until either a cached shape or input Node is encountered.
+ * @return vector of SizeType.
+ */
+template <typename TensorType>
+Shape Node<TensorType>::BatchOutputShape()
+{
+  // Return cached shape result if available.
+  Shape const candidate = op_ptr_->BatchOutputShape();
+  bool const  op_is_subgraph =
+      (std::dynamic_pointer_cast<std::shared_ptr<SubGraph<TensorType>>>(op_ptr_) != nullptr);
+
+  if (!candidate.empty() && !op_is_subgraph)
+  {
+    if (input_nodes_.empty() &&
+        (OperationType() == OpType::OP_PLACEHOLDER || OperationType() == OpType::OP_WEIGHTS))
+    {
+      FETCH_LOG_INFO(name_.c_str(), "Shape deduction reached an leaf node : " + this->name_ + " " +
+                                        OutputShapeAsString(candidate));
+    }
+    return candidate;
+  }
+
+  // If no input nodes exist - it is impossible to infer/deduce their shapes.
+  if (input_nodes_.empty())
+  {
+    FETCH_LOG_INFO(name_.c_str(), "Shape deduction reached a Graph leaf : " + this->name_);
+    return candidate;
+  }
```

Comment:
I would suggest to rearrange branching order and put check for emptiness first (because you actually called it more than once which is not necessary IMHO) it may improve the readability

Reply:
done!